fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/bitmap/ComposedBitmapIndexAggregate.java

```diff
@@ -319,9 +321,37 @@ private static boolean separateGroupFilters(@Nonnull QueryComponent filter,
                 return Optional.empty();
             }
             // Splice the index's key between the common grouping key and the position.
-            GroupingKeyExpression groupKeyExpression = (GroupingKeyExpression)indexAggregateFunction.getOperand();
-            GroupingKeyExpression fullKey = Key.Expressions.concat(groupKeyExpression.getGroupingSubKey(), indexKey, groupKeyExpression.getGroupedSubKey())
-                    .group(groupKeyExpression.getGroupedCount());
+            // The simplest place is directly before the position.
+            // But if part of the group is a nested concat, breaking that up would need support in QueryToKeyMatcher.
+            // Moreover, the caller needs to have arranged for a compatible index to exist, which requires new support to define.
+            // (https://github.com/FoundationDB/fdb-record-layer/issues/1056)
+            final GroupingKeyExpression groupKey = (GroupingKeyExpression)indexAggregateFunction.getOperand();
+            final int groupedCount = groupKey.getGroupedCount();
+            final int wholeCount = groupKey.getColumnSize();
+            int afterSpliceCount = groupedCount;
+            if (groupKey.getWholeKey() instanceof ThenKeyExpression) {
+                final List<KeyExpression> thenChildren = ((ThenKeyExpression)groupKey.getWholeKey()).getChildren();
+                int childPosition = thenChildren.size();
+                // Compute the minimum number that includes all grouped fields and keeps involved children intact.
+                afterSpliceCount = 0;
+                while (afterSpliceCount < groupedCount) {
+                    afterSpliceCount += thenChildren.get(--childPosition).getColumnSize();
+                }
+            }
+            final ThenKeyExpression splicedKey;
+            if (afterSpliceCount == groupedCount) {
+                // Preferred position at end of grouping keys.
+                splicedKey = Key.Expressions.concat(groupKey.getGroupingSubKey(), indexKey, groupKey.getGroupedSubKey());
+            } else {
+                final KeyExpression wholeKey = groupKey.getWholeKey();
+                final int splicePoint = wholeCount - afterSpliceCount;
+                if (splicePoint == 0) {
+                    splicedKey = Key.Expressions.concat(indexKey, wholeKey);
+                } else {
+                    splicedKey = Key.Expressions.concat(wholeKey.getSubKey(0, splicePoint), indexKey, wholeKey.getSubKey(splicePoint, wholeCount));
+                }
+            }
+            GroupingKeyExpression fullKey = splicedKey.group(groupedCount);
```

Comment:
Okay, it took me a while to refamiliarize myself with this code, but I think I understand now the basic approach here.

This part of the code is essentially trying out different candidate grouping expressions to match each predicate and seeing if a particular index matching that expression exists. So the important invariant is that the grouped key remains the same (and this does that), and then the grouping keys can be changed kind of at will. So, I think I understand how this works, I think.

I guess the remaining question is that it seems like the grouping key order shouldn't really matter (at least for the keys with equality predicates), and that we "should" be able to just place the `indexKey` at the beginning of the expression. But I'd also buy that the planner isn't quite sophisticated enough for that (yet?).

Reply:
Yes, given some more code here and a somewhat smarter planner, there is no fundamental reason it could not hunt around for an index some some permutation of the equality constraints. But today it just opts for a workable canonical form, which is as far to the right as possible without splitting.